### PR TITLE
Exposing transaction manager APIs to truncate invalid list through Router

### PR DESCRIPTION
**Pull request by poornachandra**
https://issues.cask.co/browse/CDAP-1540

Also updating tephra version to 0.3.5-SNAPSHOT for now. Will replace the version when tephra release is done.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java [hunk 1]
```diff
@@ -172,9 +172,27 @@ public void invalidateTx(HttpRequest request, HttpResponder responder,
                                         responder, txId);
   }
 
-  /**
-   * Reset the state of the transaction manager.
-   */
+  @Path("/transactions/invalid/remove/until")
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: are these needed in the v2/ APIs?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/TransactionHttpHandler.java [hunk 2]
```diff
@@ -110,6 +118,70 @@ public void invalidateTx(HttpRequest request, HttpResponder responder,
...
+  
+  @Path("/transactions/invalid/remove/until")
+  @POST
+  public void truncateInvalidTxBefore(HttpRequest request, HttpResponder responder) {
+    try {
+      Map<String, Long> body;
+      try {
+        body = parseBody(request, STRING_LONG_MAP_TYPE);
+      } catch (IOException e) {
+        responder.sendStatus(HttpResponseStatus.BAD_REQUEST);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/TransactionHttpHandler.java]: How about adding a message to the response?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java [hunk 3]
```diff
@@ -144,11 +144,15 @@ public AbstractAppFabricHttpHandler(Authenticator authenticator) {
   }
 
   protected int getInstances(HttpRequest request) throws IOException, NumberFormatException {
-    return parseBody(request, Instances.class).getInstances();
+    Instances instances = parseBody(request, Instances.class);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java]: this looks like an unrelated change. I have noticed this warning myself and was tempted to fix it also...
If you want to fix this, you should throw an exception that can be recognized by all callers, so that they can return BAD_REQUEST. And make that change in all callers, too :)

**Comment by anew** [comment]
I have only minor comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java (hunk 3 above)
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java]: Instances can be null when `request.getContent()` is not readable. It is may not be due to user error. We can catch `NumberFormatException` or `JsonSyntaxException` in callers and send BAD_REQUEST. I will add that too.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java (hunk 1 above)
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: We need these for 2.6.2.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/TransactionHttpHandler.java (hunk 2 above)
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/TransactionHttpHandler.java]: Good point.

**Comment by poornachandra** [comment]
@anew Addressed comments, please take a look.

**Comment by anew** [comment]
LGTM

**Comment by poornachandra** [comment]
Build is green http://builds.cask.co/browse/CDAP-DUT1056-8
